File: src/embed/ollama.ts
import { Ollama } from 'ollama';
import { config } from '../config.js';
import { logger } from '../logger.js';

const client = new Ollama({ host: config.OLLAMA_HOST });

export async function embedOllama(texts: string[], batchSize = 16): Promise<number[][]> {
  const result: number[][] = [];
  for (let i = 0; i < texts.length; i += batchSize) {
    const batch = texts.slice(i, i + batchSize);
    const res = await client.embed({
      model: config.OLLAMA_EMBED_MODEL,
      input: batch,
      truncate: true,
    });
    if (!res.embeddings || res.embeddings.length !== batch.length) throw new Error('Unexpected response');
    for (const v of res.embeddings) {
      if (v.length !== config.EMBED_DIM) throw new Error(`Dim mismatch: ${v.length}`);
      result.push(v);
    }
    logger.debug({ done: result.length, total: texts.length }, 'embed (ollama)');
  }
  return result;
}

File: crates/embed/src/ollama.rs
use rag_common::{AppError, Config, Result};
use serde::Deserialize;
use serde_json::json;

const BATCH: usize = 16;

#[derive(Debug, Deserialize)]
struct EmbedResponse {
    embeddings: Vec<Vec<f32>>,
}

pub async fn embed(texts: &[String]) -> Result<Vec<Vec<f32>>> {
    let cfg = Config::get();
    let client = reqwest::Client::new();
    let url = format!("{}/api/embed", cfg.ollama_host.trim_end_matches('/'));

    let mut out: Vec<Vec<f32>> = Vec::with_capacity(texts.len());
    for chunk in texts.chunks(BATCH) {
        let body = json!({
            "model": cfg.ollama_embed_model,
            "input": chunk,
            "truncate": true,
        });
        let resp = client
            .post(&url)
            .json(&body)
            .send()
            .await?
            .error_for_status()
            .map_err(AppError::http)?;
        let parsed: EmbedResponse = resp.json().await?;
        for v in parsed.embeddings {
            if v.len() != cfg.embed_dim as usize {
                return Err(AppError::DimMismatch {
                    expected: cfg.embed_dim as usize,
                    got: v.len(),
                });
            }
            out.push(v);
        }
    }
    Ok(out)
}
